public/app/features/dashboard/state/runRequest.ts

```diff
@@ -226,3 +227,34 @@ export function preProcessPanelData(data: PanelData, lastResult: PanelData): Pan
     timings: { dataProcessingTime: STOPTIME - STARTTIME },
   };
 }
+
+export function preProcessAppendPanelData(
+  data: PanelData,
+  lastResult: PanelData,
+  direction: QueryDirection
+): PanelData {
+  const { series } = data;
+
+  //  for loading states with no data, use last result
+  if (data.state === LoadingState.Loading && series.length === 0) {
+    if (!lastResult) {
+      lastResult = data;
+    }
+
+    return { ...lastResult, state: LoadingState.Loading };
+  }
+
+  // Make sure the data frames are properly formatted
+  const STARTTIME = performance.now();
+  const dataFrames =
+    direction === QueryDirection.backward ? [...lastResult.series, ...series] : [...series, ...lastResult.series];
```

Comment:
Appends DataFrames on the very beginning or the very end of data (DataFrames) array to make drops slightly more straightforward.